nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+                         .format(numerical_grads))
+
+    input_vars = [x for x in symbol.list_input_variables()]
+    input_dict = {x.attr('name'): x for x in input_vars}
+
+    if grad_input_vars is None:
+        grad_input_vars = input_vars
+
+    shape = shape.copy() if shape else {}
+
+    grad_input_vars_real = []
+    for x in grad_input_vars:
+        if isinstance(x, tuple):
+            if len(x) != 2:
+                raise ValueError("Expected (var_name, shape), not {}".format(x))
+            x_name = x[0] if isinstance(x[0], str) else x[0].attr('name')
+            if x_name in shape:
+                if shape[x_name] != x[1]:
+                    raise ValueError("Can't decide which shape to use for {}: "
+                                     "{} from shape or {} from grad_input_vars"
+                                     .format(x_name, shape[x_name], x[1]))
+            shape[x_name] = x[1]
+            x = x[0]
+
+        grad_input_vars_real.append(input_dict[x] if isinstance(x, str) else x)
+
+    grad_input_vars = grad_input_vars_real
+
+    # Infer the output shape and dtype by creating a graph and running passes
+
+    forward_graph = nnvm.graph.create(symbol)
+
+    if dtype is not None:
+        nnvm.compiler.graph_attr.set_dtype_inputs(forward_graph, dtype)
+
+    if shape is not None:
+        nnvm.compiler.graph_attr.set_shape_inputs(forward_graph, shape)
+
+    forward_graph = forward_graph.apply('InferShape').apply('InferType')
+    shapes = forward_graph.json_attr('shape')
+    dtypes = forward_graph.json_attr('dtype')
+    out_shape = shapes[forward_graph.index.output_entries[0][0]]
+    out_dtype = dtypes[forward_graph.index.output_entries[0][0]]
+
+    backward_graph = None
+
+    # If we want gradients, we have to recreate the graph, but now with gradient computations
+    # Note that here we need out_shape for defining the shape of head grads, so we have to
+    # create the graph twice
+    if np_backward is not None or numerical_grads:
+        try:
+            head_grads_symbol = nnvm.symbol.Variable("head_grads", shape=out_shape, dtype=out_dtype)
+            grad_symbols = graph_util.gradients([symbol], grad_input_vars,
+                                                grad_ys=head_grads_symbol)
+            # Sometimes grads do not depend on head_grads, so head_grads does not appear
+            # in the variable list; adding it manually prevents this, making things a bit easier
+            backward_graph = \
+                nnvm.graph.create(nnvm.symbol.Group([symbol] + grad_symbols + [head_grads_symbol]))
+
+            if dtype is not None:
+                nnvm.compiler.graph_attr.set_dtype_inputs(backward_graph, dtype)
+
+            if shape is not None:
+                nnvm.compiler.graph_attr.set_shape_inputs(backward_graph, shape)
+
+            backward_graph = backward_graph.apply('InferShape').apply('InferType')
```

Comment:
With graph.apply("Gradient"), we don't need to infershape/dtype again.

Reply:
https://github.com/dmlc/tvm/blob/master/nnvm/tests/python/compiler/test_top_level1.py#L32-L37
We just need to update the shape head_grads for input shape dictionary when compiling.

Reply:
Well, I perform shape and type inference here for two reasons: it makes it easier to query shape and dtype information during random input generation, and it is nice to have a graph printed out with inferred shapes and dtypes when the function fails. Of course, it is possible to avoid the second shape and dtype inference by manually adding head_grads to the list of variables we want to generate data for, but I see no benefits.

Reply:
OK. This looks good to me.